src/cmd/vendor.rs

```diff
@@ -0,0 +1,645 @@
+                    })))
+                    .and_then(|_| async {
+                        let rev_hash = match vendor_package.upstream {
+                            config::Dependency::GitRevision(_, ref rev) => rev,
+                            _ => unimplemented!(),
+                        };
+                        if *rev_hash != git.spawn_with(|c| c.arg("rev-parse").arg("--verify").arg(format!("{}^{{commit}}", rev_hash))).await?.trim_end_matches('\n') {
+                            Err(Error::new("Please ensure your vendor reference is a commit hash to avoid upstream changes impacting your checkout"))
+                        } else {
+                            Ok(())
+                        }
+                    })
+                    .await
+                })?;
+
+                tmp_path.to_path_buf()
+            }
+            DependencySource::Registry => unimplemented!(),
+        };
+
+        // Extract patch dirs of links
+        let mut patch_links: Vec<PatchLink> = Vec::new();
+        for link in vendor_package.mapping.clone() {
+            patch_links.push(PatchLink {
+                patch_dir: link.patch_dir,
+                from_prefix: link.from,
+                to_prefix: link.to,
+            })
+        }
+
+        // If links do not specify patch dirs, use package-wide patch dir
+        let patch_links = {
+            match patch_links[..] {
+                [] => vec![PatchLink {
+                    patch_dir: vendor_package.patch_dir.clone(),
+                    from_prefix: PathBuf::from(""),
+                    to_prefix: PathBuf::from(""),
+                }],
+                _ => patch_links,
+            }
+        };
+
+        let git = Git::new(tmp_path, &sess.config.git);
+
+        match matches.subcommand() {
+            Some(("diff", _)) => {
+                // Apply patches
+                patch_links.clone().into_iter().try_for_each(|patch_link| {
+                    apply_patches(&rt, git, vendor_package.name.clone(), patch_link).map(|_| ())
+                })?;
+
+                // Stage applied patches to clean working tree
+                rt.block_on(git.add_all())?;
+
+                // Print diff for each link
+                patch_links.into_iter().try_for_each(|patch_link| {
+                    let get_diff = diff(&rt, git, vendor_package, patch_link, dep_path.clone())
+                        .map_err(|cause| Error::chain("Failed to get diff.", cause))?;
+                    if !get_diff.is_empty() {
+                        print!("{}", get_diff);
+                    }
+                    Ok(())
+                })
+            }
+
+            Some(("init", matches)) => {
+                patch_links.clone().into_iter().try_for_each(|patch_link| {
+                    stageln!("Copying", "{} files from upstream", vendor_package.name);
+                    // Remove existing directories before importing them again
+                    std::fs::remove_dir_all(
```

Comment:
wrap this with std::fs::try_exists and if it doesn't, create with std::fs::create_dir_all. Also we should explicitly handle the None case. Just using unwrap is kind of an anti-pattern I heard and the corresponding panik message is often not very useful for the user to figure out what was going wrong.